src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp

```diff
@@ -218,10 +218,65 @@ size_t ShenandoahOldGeneration::unexpend_promoted(size_t decrement) {
   return Atomic::sub(&_promoted_expended, decrement);
 }
 
-size_t ShenandoahOldGeneration::get_promoted_expended() {
+size_t ShenandoahOldGeneration::get_promoted_expended() const {
   return Atomic::load(&_promoted_expended);
 }
 
+bool ShenandoahOldGeneration::can_allocate(const ShenandoahAllocRequest &req) const {
+  assert(req.type() != ShenandoahAllocRequest::_alloc_gclab, "GCLAB pertains only to young-gen memory");
+
+  const size_t requested_bytes = req.size() * HeapWordSize;
+  if (can_promote(requested_bytes)) {
+    // The promotion reserve should be able to accommodate this request. The request
+    // might still fail if alignment with the card table increases the size. The request
+    // may also fail if the heap is badly fragmented and the free set cannot find room for it.
+    return true;
+  }
+
+  if (req.type() == ShenandoahAllocRequest::_alloc_plab) {
+    // The promotion reserve is exhausted. Check if we still have room for evacuations.
+    // Note that we cannot really know how much of the plab will be used for evacuations,
+    // so here we only check that some evacuation reserve still exists.
+    return get_evacuation_reserve() > 0;
+  }
+
+  // This is a shared allocation promotion request. However, we do not have room for any
```

Comment:
How do we know this is NOT a shared allocation evacuation request?  Is there a precondition on the method call?

Implementation below looks we're not sure this is a promotion.  We still need to check.

I think what we're saying here is:
  if req.is_promotion();   return false because we already checked above, and found that can_promote(requested_bytes) to be false.
  
  So what if req is not promotion?  wouldn't we want to confirm that get_evacuation_reserve() > requested_bytes?

Reply:
> How do we know this is NOT a shared allocation evacuation request?
We check if the request type is  `ShenandoahAllocRequest::_alloc_plab` on line 239.

> So what if req is not promotion? wouldn't we want to confirm that get_evacuation_reserve() > requested_bytes?
[The previous version of this code](https://github.com/openjdk/shenandoah/pull/427/files#diff-5ca0a05384b7b2604dd3c9b55d91a7010a42cd0ba246600c844e19821bd6b60bL1178) had a comment for this case (i.e., shared evacuation) that read:
```C
// This is a shared allocation for evacuation. Memory has already been reserved for this purpose.
```
https://github.com/openjdk/shenandoah/pull/427/files#diff-5ca0a05384b7b2604dd3c9b55d91a7010a42cd0ba246600c844e19821bd6b60bL1178

I believe this change preserves the original behavior, but it makes sense to me to also enforce evacuation reserve constraints for `shared` evacuations.

Reply:
I've fixed these comments.